atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java

```diff
@@ -2330,30 +2342,36 @@ protected Map<Cell, Long> detectWriteAlreadyCommittedInternal(
     private boolean rollbackFailedTransactions(
             TableReference tableRef,
             Map<Cell, Long> keysToDelete,
-            Map<Long, Long> commitTimestamps,
+            LongLongMap commitTimestamps,
             TransactionService transactionService) {
-        for (long startTs : new HashSet<>(keysToDelete.values())) {
-            if (commitTimestamps.get(startTs) == null) {
-                log.warn("Rolling back transaction: {}", SafeArg.of("startTs", startTs));
-                if (!rollbackOtherTransaction(startTs, transactionService)) {
-                    return false;
+        ImmutableLongSet timestamps = LongSets.immutable.ofAll(keysToDelete.values());
+        boolean allRolledBack = timestamps.allSatisfy(startTs -> {
+            if (commitTimestamps.containsKey(startTs)) {
+                long commitTs = commitTimestamps.get(startTs);
+                if (commitTs != TransactionConstants.FAILED_COMMIT_TS) {
+                    throw new SafeIllegalArgumentException(
+                            "Cannot rollback already committed transaction",
+                            SafeArg.of("startTs", startTs),
+                            SafeArg.of("commitTs", commitTs));
                 }
-            } else {
-                Preconditions.checkArgument(commitTimestamps.get(startTs) == TransactionConstants.FAILED_COMMIT_TS);
             }
-        }
+            log.warn("Rolling back transaction: {}", SafeArg.of("startTs", startTs));
```

Comment:
I think we only want to do this if `startTs` is not in `commitTimestamps` - so this and the rollbackOther needs to be in an else block; or the end of the if-block above should have a `return true;`?